activerecord/test/cases/transaction_callbacks_test.rb

```diff
@@ -108,6 +108,19 @@ def test_call_after_commit_after_transaction_commits
     assert_equal [:after_commit], @first.history
   end
 
+  def test_do_not_call_after_commit_or_after_update_commit_after_transaction_commits_if_update_fails
+    @first.after_commit_block { |r| r.history << :after_commit }
+    @first.after_commit_block(:update) { |r| r.history << :commit_on_update }
+    def @first.valid?(*args)
+      false
+    end
+    saved = TopicWithCallbacks.transaction do
+      @first.save
+    end
+    assert_equal false, saved
```

Comment:
FYI: This assertion is here for the benefit of the reader. It demonstrates that the save actually failed, which is why we want the callback _not_ to be called.  I can replace it with a comment to that effect if that would be preferable.